### PR TITLE
Dockerize Iglu Server

### PR DESCRIPTION
**Pull request by oguzhanunlu**
This PR aligns work from #11 per our docker approach (a folder with Dockerfile per version).

Iglu Registry `0.3.0` isn't published yet but this PR includes Dockerfile for that too.

### CODE REVIEW

File: iglu-registry/0.3.0/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+RUN apk update && \
+    apk add postgresql-client
+
+COPY ./scripts/super_api_key_init.sh ${SNOWPLOW_BIN_PATH}
+RUN chmod +x ${SNOWPLOW_BIN_PATH}/super_api_key_init.sh && \
+	sed -i -e 's/\r$//' ${SNOWPLOW_BIN_PATH}/super_api_key_init.sh
+
+WORKDIR $SNOWPLOW_BIN_PATH
+
+EXPOSE 8080
```
Comment (BenFradet) [inline iglu-registry/0.3.0/Dockerfile]: I'd rather specify the entrypoint here than in the docker compose example

File: iglu-registry/README.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+```
+$ docker-compose up -d
+```
+
+### Insert the super API key
+
+The following command inserts the super API key which can be used for generating API keys with read/write privileges for a specific vendor prefix.
+
+```
+ $ docker exec iglu-registry /snowplow/bin/super_api_key_init.sh
```
Comment (BenFradet) [inline iglu-registry/README.md]: why does the user need to do this? can't this be part of the container setup

**Review by BenFradet — COMMENTED** [review]
It seems a lot of weight rests on the shoulders of the user , I'd rather we alleviate that weight.

You can have a look at the existing images.

Also the example might be better off in its own folder.

**Comment by BenFradet** [comment]
also why do we need 0.2.0?

**Comment by oguzhanunlu** [comment]
Thanks for feedback @BenFradet ! I included `0.2.0` since other apps have their previous versions dockerized. Should we keep only latest version of iglu registry?

**Comment by BenFradet** [comment]
since we've released this project we've been incrementally adding images but not retroactively. so I don't think it's useful to add 0.2.0

File: iglu-registry/0.3.0/Dockerfile (hunk 1 above)
Comment (oguzhanunlu) [reply iglu-registry/0.3.0/Dockerfile]: Entrypoint includes `-Dconfig.file=<config_file>` which contains db configuration. Moving entrypoint here requires there is a postgres instance with that db credentials. Otherwise iglu registry will throw MissingConfiguration exception and stop. In short, running iglu-registry without postgres is useless. Should I still include entrypoint here @BenFradet ?

File: iglu-registry/README.md (hunk 2 above)
Comment (oguzhanunlu) [reply iglu-registry/README.md]: This script inserts super API key, provided in `docker-compose.yml` , to `apikeys` table in postgres. I don't think iglu-registry container should run this command without being sure about postgres, which docker-compose provides.

**Comment by oguzhanunlu** [comment]
Oh, alright then. I'll remove `0.2.0`.

File: iglu-registry/0.3.0/Dockerfile (hunk 1 above)
Comment (BenFradet) [reply iglu-registry/0.3.0/Dockerfile]: yes, you need an entrypoint so you can control how the app is launched

and no you don't need to specify the config file in the entrypoint

as I said before, cf other docker images

Comment (oguzhanunlu) [reply iglu-registry/0.3.0/Dockerfile]: Ah, right, I was always overriding default config.

Comment (oguzhanunlu) [reply iglu-registry/0.3.0/Dockerfile]: Hey @chuwy there are two concerns we need to decide on about dockerizing iglu-registry. They came out after moving `ENTRYPOINT` from docker-compose to Dockerfile, per @BenFradet 's feedback.

1) Would that be possible to add another issue to r9 for adding `--config` CLI parameter to iglu-registry? Currently only way to provide custom config is through system property, `-Dconfig.file=<path_to_config_file>` and this config approach disables us to make real use of `CMD`  instruction when writing Dockerfile. Other dockerized apps have a `--config` CLI parameter and people can provide custom config file by mounting a config directory and providing this `--config` option.

2) If we are positive for first question, can we also add `--help` CLI parameter to iglu-registry?

@BenFradet 's suggestion is to have a uniform config approach with our apps.

Comment (alexanderdean) [reply iglu-registry/0.3.0/Dockerfile]: I like this approach

Comment (chuwy) [reply iglu-registry/0.3.0/Dockerfile]: @oguzhanunlu sure, it makes sense!

**Comment by oguzhanunlu** [comment]
Hey @BenFradet how does it look now?

File: iglu-registry/0.3.0/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    mv ${SNOWPLOW_BIN_PATH}/iglu-server-* ${SNOWPLOW_BIN_PATH}/iglu_server.jar && \
+    chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
+
+# Install postgresql
+RUN apk update && \
+    apk add postgresql-client
+
+COPY ./scripts/super_api_key_init.sh ${SNOWPLOW_BIN_PATH}
+RUN chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
+RUN chmod +x ${SNOWPLOW_BIN_PATH}/super_api_key_init.sh && \
```
Comment (BenFradet) [inline iglu-registry/0.3.0/Dockerfile]: no need for an additional layer here

File: iglu-registry/0.3.0/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    rm -rf /tmp/build && \
+    mv ${SNOWPLOW_BIN_PATH}/iglu-server-* ${SNOWPLOW_BIN_PATH}/iglu_server.jar && \
+    chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
+
+# Install postgresql
+RUN apk update && \
+    apk add postgresql-client
+
+COPY ./scripts/super_api_key_init.sh ${SNOWPLOW_BIN_PATH}
+RUN chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
```
Comment (BenFradet) [inline iglu-registry/0.3.0/Dockerfile]: this is already done above

File: iglu-registry/0.3.0/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ENV ARCHIVE="iglu_server_${IGLU_REGISTRY_VERSION}.zip"
+
+# Install Iglu Registry
+RUN mkdir -p /tmp/build && \
+    cd /tmp/build && \
+    wget -q http://dl.bintray.com/snowplow/snowplow-generic/${ARCHIVE} && \
+    unzip -d ${SNOWPLOW_BIN_PATH} ${ARCHIVE} && \
+    cd /tmp && \
+    rm -rf /tmp/build && \
+    mv ${SNOWPLOW_BIN_PATH}/iglu-server-* ${SNOWPLOW_BIN_PATH}/iglu_server.jar && \
```
Comment (BenFradet) [inline iglu-registry/0.3.0/Dockerfile]: why the rename?

File: iglu-registry/0.3.0/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
+
+# Install postgresql
+RUN apk update && \
+    apk add postgresql-client
+
+COPY ./scripts/super_api_key_init.sh ${SNOWPLOW_BIN_PATH}
+RUN chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
+RUN chmod +x ${SNOWPLOW_BIN_PATH}/super_api_key_init.sh && \
+	sed -i -e 's/\r$//' ${SNOWPLOW_BIN_PATH}/super_api_key_init.sh
```
Comment (BenFradet) [inline iglu-registry/0.3.0/Dockerfile]: why is this necessary?

File: iglu-registry/README.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+This folder contains
+
+* docker images of [Iglu Registry](iglu-registry)
+  * `0.3.0`
+* a docker compose example to run [Iglu Registry](iglu-registry) with [Postgres](https://github.com/docker-library/postgres)
+
+[Iglu Registry](iglu-registry) images are based on [Snowplow base image][base-image] which leverages [the Java 8 Alpine image][alpine-image].
+
+The S3 Loader runs under [dumb-init][dumb-init] which handles reaping zombie processes
```
Comment (BenFradet) [inline iglu-registry/README.md]: s3 loader

File: iglu-registry/README.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+and forwards signals on to all processes running in the container. This image also uses
+[su-exec][su-exec], as a sudo replacement, to run the S3 Loader as the non-root `snowplow` user.
+
+The container exposes the `/snowplow/config` volume to store the loader configuration. If this
+folder is bind mounted then ownership will be changed to the `snowplow` user.
+
+The `-XX:+UnlockExperimentalVMOptions` and `-XX:+UseCGroupMemoryLimitForHeap` JVM options will be
+automatically provided when launching the loader in order to make the JVM adhere to the memory
```
Comment (BenFradet) [inline iglu-registry/README.md]: loader

File: iglu-registry/0.3.0/scripts/super_api_key_init.sh [hunk 9]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+pswd=$(grep password /snowplow/config/application.conf | cut -d '=' -f 2 | tr -d '"' | tr -d '\r')
+host=$(grep host /snowplow/config/application.conf | cut -d '=' -f 2 | tr -d '"' | tr -d '\r' | tail -n1)
+port=$(grep port /snowplow/config/application.conf | cut -d '=' -f 2 | tr -d '"' | tr -d '\r' | tail -n1)
+dbname=$(grep dbname /snowplow/config/application.conf | cut -d '=' -f 2 | tr -d '"' | tr -d '\r')
+username=$(grep username /snowplow/config/application.conf | cut -d '=' -f 2 | tr -d '"' | tr -d '\r')
+PGPASSWORD=$(echo $pswd) psql -h $(echo $host) -p $(echo $port) -d $(echo $dbname) -U $(echo $username) -c "insert into apikeys (uid, vendor_prefix, permission, createdat) values ($(echo "'"$API_KEY"'"), '.', 'super', current_timestamp);"
```
Comment (BenFradet) [inline iglu-registry/0.3.0/scripts/super_api_key_init.sh]: since we're making the user launch this script, why not make the user directly launch the psql command?

Also, @jbeemster suggested that maybe we should directly specify the super api key as a secret in the config and the table bootstrapping would take care of inserting it directly in the table, which I think is a good idea.
We would need to find a way to store secrets in the config though.

Wdyt @chuwy @oguzhanunlu ?

File: iglu-registry/example/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        ports:
+         - "8080:8080"
+        volumes:
+         - "./config/application.conf:/snowplow/config/application.conf"
+        logging:
+          options:
+            max-size: 1G
+        environment:
+          - API_KEY=1d9c7e70-012b-11e8-ba89-0ed5f89f718b
+    postgres:
```
Comment (BenFradet) [inline iglu-registry/example/docker-compose.yml]: Can we put postgres first since iglu-registry depends on it?

File: iglu-registry/0.3.0/scripts/super_api_key_init.sh (hunk 9 above)
Comment (BenFradet) [reply iglu-registry/0.3.0/scripts/super_api_key_init.sh]: ah snowplow/iglu#63

File: iglu-registry/README.md (hunk 7 above)
Comment (oguzhanunlu) [reply iglu-registry/README.md]: Ah, copy-pasting common stuff, will fix it.

File: iglu-registry/example/docker-compose.yml (hunk 10 above)
Comment (oguzhanunlu) [reply iglu-registry/example/docker-compose.yml]: Sure.

File: iglu-registry/0.3.0/Dockerfile (hunk 4 above)
Comment (oguzhanunlu) [reply iglu-registry/0.3.0/Dockerfile]: Do you mean base image referring to "above" @BenFradet ? This is here because I noticed that, after `COPY` command, user and group permissions of `super_api_key_init.sh` is `root:root`, not `snowplow:snowplow`.

File: iglu-registry/0.3.0/Dockerfile (hunk 5 above)
Comment (oguzhanunlu) [reply iglu-registry/0.3.0/Dockerfile]: No significant reason. Will remove it.

File: iglu-registry/0.3.0/Dockerfile (hunk 4 above)
Comment (BenFradet) [reply iglu-registry/0.3.0/Dockerfile]: the chown is done twice

Comment (BenFradet) [reply iglu-registry/0.3.0/Dockerfile]: in your dockerfile

File: iglu-registry/0.3.0/Dockerfile (hunk 3 above)
Comment (oguzhanunlu) [reply iglu-registry/0.3.0/Dockerfile]: Right, it has already executable bit set in git index. Will remove it.

Comment (BenFradet) [reply iglu-registry/0.3.0/Dockerfile]: you should try to minimize the number of `RUN` and `COPY` that's what I meant

File: iglu-registry/0.3.0/Dockerfile (hunk 4 above)
Comment (oguzhanunlu) [reply iglu-registry/0.3.0/Dockerfile]: Oh sorry, I noticed now. Will remove first one.

File: iglu-registry/0.3.0/Dockerfile (hunk 3 above)
Comment (oguzhanunlu) [reply iglu-registry/0.3.0/Dockerfile]: I see. I could chain commands.

File: iglu-registry/0.3.0/scripts/super_api_key_init.sh (hunk 9 above)
Comment (BenFradet) [reply iglu-registry/0.3.0/scripts/super_api_key_init.sh]: so could we remove this and make the user launch the psql command directly at least?

Comment (oguzhanunlu) [reply iglu-registry/0.3.0/scripts/super_api_key_init.sh]: Sure, [this command](https://github.com/snowplow/snowplow-docker/blob/iglu-registry/iglu-registry/README.md#insert-the-super-api-key) will be changed to something like `docker exec iglu-registry psql ...` instead of `docker exec iglu-registry sh ...`.

**Comment by chuwy** [comment]
Hey @oguzhanunlu, @BenFradet as a side-note: do you think we should add Iglu Registry to README?

File: iglu-registry/README.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+    `$ git clone git@github.com:snowplow/snowplow-docker.git`
+
+2) Navigate to `iglu-registry/`
+
+    `$ cd iglu-registry/`
+
+3) Build image    
+
+    `$ docker build -t iglu-registry:0.x.0 iglu-registry/0.x.0`
```
Comment (chuwy) [inline iglu-registry/README.md]: Right command here is `docker build -t iglu-registry:0.x.0 0.x.0`

File: iglu-registry/README.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    where `0.x.0` is version of Iglu Registry one wants to run. Each version of Iglu Registry has a dedicated folder under `iglu-registry/` which contains a dedicated `Dockerfile`.
+
+### Usage
+
+After building an image for registry, there are two options:
+
+  1) Running standalone image with postgres on docker host
+  2) Running with postgres in docker-compose
+
+Either way, a super API key needs to be inserted into database. `Iglu Registry` makes use of an UUID key, called super API key, to generate other API keys.
```
Comment (chuwy) [inline iglu-registry/README.md]: Unnecessary comma in `called super API key, to generate`

File: iglu-registry/README.md (hunk 11 above)
Comment (oguzhanunlu) [reply iglu-registry/README.md]: Ah, right. Will update.

File: iglu-server/README.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+  --version            Print version info
+  --config <filename>  Path to custom config file. It is required and can not be empty.
+```
+
+Alternatively, we can mount a configuration folder and run the server:
+
+```bash
+$ docker run \
+  -d \
+  -v ${PWD}/config:/snowplow/config \
```
Comment (chuwy) [inline iglu-server/README.md:86]: We're still in `snowplow-docker/iglu-registry` (according to docs), so `snowplow-docker/iglu-registry/config` won't work. We either need to add `cd 0.x.0` somewhere or specify full paths.

Comment (chuwy) [reply iglu-server/README.md:86]: I prefer first option.

**Comment by oguzhanunlu** [comment]
@BenFradet I addressed your last feedback on super api key insertion and @chuwy 's feedback on both READMEs.  How does it look? Especially, updated repository README and [insert super api key method](https://github.com/snowplow/snowplow-docker/blob/iglu-registry/iglu-server/README.md#insert-the-super-api-key)

Temp commit is for QA. After we are done with QA, I'll remove it.

File: iglu-server/README.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+  -e 'SP_JAVA_OPTS=-Xms512m -Xmx512m' \
+  iglu-server:0.x.0 \
+  --config /snowplow/config/application.conf
+```
+
+#### 2. Running with postgres in docker-compose
+
+An example docker-compose for iglu server and postgres can be found at [the docker compose example][docker-compose-example].
+
+Before running, make sure to replace the API_KEY environment variable in `iglu-server/example/docker-compose.yml` with your preferred super API key. e.g. 1d9c7e70-012b-11e8-ba89-0ed5f89f718b
```
Comment (chuwy) [inline iglu-server/README.md]: I'm wondering if we should provide default key (this one for example). Removing one more manual step. Those who concerned about security will change it anyway. WDYT @alexanderdean, @oguzhanunlu, @BenFradet?

Comment (BenFradet) [reply iglu-server/README.md]: yes this has already been discussed earlier in the reivew, c.f. snowplow/iglu#63

File: iglu-server/README.md (hunk 13 above)
Comment (BenFradet) [reply iglu-server/README.md:86]: this is an example command, there is no config and there shouldn't be, except in the docker compose example of course.

File: iglu-server/example/docker-compose.yml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+         - postgres
+        ports:
+         - "8080:8080"
+        volumes:
+         - "./config/application.conf:/snowplow/config/application.conf"
+        logging:
+          options:
+            max-size: 1G
+        environment:
+          - API_KEY=1d9c7e70-012b-11e8-ba89-0ed5f89f718b
```
Comment (BenFradet) [inline iglu-server/example/docker-compose.yml]: is this still needed btw?

File: iglu-server/README.md (hunk 14 above)
Comment (BenFradet) [reply iglu-server/README.md]: I don't think this part is still needed anyway given the `Insert the super API key` section below

File: iglu-server/README.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+Now you are ready to use Iglu Server. Below are sample usages for some endpoints.
+
+### Generate read/write API keys for a vendor
+
+The following POST request generates read and write API keys for a specific vendor prefix returned in JSON format.
+
+```
+curl -X POST \
+  http://localhost:8080/api/auth/keygen?vendor_prefix=<a_vendor_prefix> \
```
Comment (chuwy) [inline iglu-server/README.md:134]: `localhost` doesn't work due CORS. Did we change something regarding that?

File: iglu-server/0.3.0/docker-entrypoint.sh [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+
+# If the config directory has been mounted through -v, we chown it.
+if [ "$(stat -c %u ${SNOWPLOW_CONFIG_PATH})" != "$(id -u snowplow)" ]; then
+  chown -R snowplow:snowplow ${SNOWPLOW_CONFIG_PATH}
+fi
+
+# Make sure we run the server as the snowplow user
+exec su-exec snowplow:snowplow /usr/bin/java \
+  $SP_JAVA_OPTS -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap \
+  -jar ${SNOWPLOW_BIN_PATH}/iglu-server-0.3.0-rc3.jar "$@"
```
Comment (BenFradet) [inline iglu-server/0.3.0/docker-entrypoint.sh]: remember to change that to use `IGLU_SERVER_VERSION` once 0.3.0 is released :+1:

Comment (oguzhanunlu) [reply iglu-server/0.3.0/docker-entrypoint.sh]: There are 2 commits and this change is inside tmp commit @BenFradet , once QA is finished, I'll remove tmp commit.

File: iglu-server/example/docker-compose.yml (hunk 15 above)
Comment (oguzhanunlu) [reply iglu-server/example/docker-compose.yml]: Again tmp commit. Needed for QA.

File: iglu-server/README.md (hunk 14 above)
Comment (oguzhanunlu) [reply iglu-server/README.md]: Then I'll make use of this example key as default and put a reminder in README for people concerned about security.

Comment (oguzhanunlu) [reply iglu-server/README.md]: @BenFradet right, no need to provide that environment variable.

**Comment by BenFradet** [comment]
also @oguzhanunlu don't forget to get the image into the travis matrix

**Comment by alexanderdean** [comment]
We are calling this Iglu Server to prevent confusion...

File: iglu-server/README.md (hunk 13 above)
Comment (chuwy) [reply iglu-server/README.md:86]: Ok, makes sense. @oguzhanunlu could you please at least link `application.conf` in first paragraph of this section then.

File: iglu-server/README.md (hunk 16 above)
Comment (oguzhanunlu) [reply iglu-server/README.md:134]: I just released `rc5` of iglu server which resolve this issue.

**Comment by oguzhanunlu** [comment]
@chuwy @BenFradet I addressed last set of feedback on both r9 PR and this PR. I published new artifact and tested both the jar itself and dockerized version. Let me know if there is something missing.

File: iglu-server/README.md [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+
+    `$ git clone git@github.com:snowplow/snowplow-docker.git`
+
+2) Navigate to `snowplow-docker/iglu-server/`
+
+    `$ cd snowplow-docker/iglu-server/`
+
+3) Build image    
+
+    `$ docker build -t iglu-server:0.x.0 0.x.0`
```
Comment (BenFradet) [inline iglu-server/README.md]: can you use $VERSION like what has been done in 67d96e75da1b1ecf292dc189b2411d3e2e108ac6?

**Review by BenFradet — APPROVED** [review]
LGTM :+1:

**Comment by oguzhanunlu** [comment]
@BenFradet I updated example `application.conf` and tested per [new config parameter](https://github.com/snowplow/iglu/issues/362) introduced for iglu server. It solves deployment issue with Swagger UI. [Related issue](https://github.com/snowplow/iglu/issues/362#issue-312568282) has a background story if you'd like to look at.

Let me know if there is anything breaking your LGTM confirmation.

**Comment by BenFradet** [comment]
>Let me know if there is anything breaking your LGTM confirmation.

nope all good for me, I just renamed the issue though

**Review by chuwy — APPROVED** [review]
Nice work, @oguzhanunlu!

**Comment by alexanderdean** [comment]
Woohoo!

**Review by alexanderdean — APPROVED** [review]

**Comment by BenFradet** [comment]
cherry-picked into #42, closing. Thanks!
